consensus/hotstuff/safetyrules/safety_rules.go

```diff
@@ -45,42 +46,53 @@ func New(
 //  * (nil, model.NoVoteError): If the voter decides that it does not want to vote for the given block.
 //    This is a sentinel error and _expected_ during normal operation.
 // All other errors are unexpected and potential symptoms of uncovered edge cases or corrupted internal state (fatal).
-func (v *SafetyRules) ProduceVote(proposal *model.Proposal, curView uint64) (*model.Vote, error) {
+func (r *SafetyRules) ProduceVote(proposal *model.Proposal, curView uint64) (*model.Vote, error) {
 	block := proposal.Block
 	// sanity checks:
 	if curView != block.View {
 		return nil, fmt.Errorf("expecting block for current view %d, but block's view is %d", curView, block.View)
 	}
-	if curView <= v.lastVotedView {
-		return nil, fmt.Errorf("current view (%d) must be larger than the last voted view (%d)", curView, v.lastVotedView)
+
+	err := r.IsSafeToVote(proposal)
+	if err != nil {
+		return nil, fmt.Errorf("not safe to vote for proposal %x: %w", proposal.Block.BlockID, err)
+	}
+
+	// we expect that only valid proposals are submitted for voting
+	// we need to make sure that proposer is not ejected to decide to vote or not
+	_, err = r.committee.IdentityByBlock(block.BlockID, block.ProposerID)
+	if model.IsInvalidSignerError(err) {
+		return nil, model.NoVoteError{Msg: "not voting - proposer ejected"}
```

Comment:
Technically from just this check we can't assume the proposer is ejected. It could also be either non-existent or have zero weight.
```suggestion
		// the proposer must be ejected since the proposal has already been validated,
		// which ensures that the proposer was a valid committee member at the start of the epoch
		return nil, model.NoVoteError{Msg: "not voting - proposer ejected"}
```